File: mtscrape.rb
#!/usr/bin/ruby
#
# mtscrape [OPTIONS] [SOURCES]
#
# Options:
#   -a, --age       Maximum age of feed item
#   -c, --convert   Convert wmv to ogg/theora
#   -d, --dir       Output directory
#   -f, --fast      Stream low-quality (faster)
#   -m, --match     Regular expression to match against items
#   -h, --help      This help text
#   -v, --verbose   Verbose output
#
# Sources:
#   -A, --asx       Parse ASX link/file directly
#   -C, --category  Parse RSS feed of mediathek category ID
#   -I, --item      Parse JSON of mediathek item ID
#   -L, --link      Parse JSON of mediathek link
#
#  Note: -a and -m are only used in combination with -C
#        -f is not used in combination with -A
#  Note: you can specify all sources multiple times to download multiple items.

require 'rubygems'
require 'date'
require 'getoptlong'
require 'json' # gem install json
require 'open-uri'
require 'rdoc/usage'
require 'rss/1.0'
require 'rss/2.0'
require 'xml/libxml' # gem install libxml-ruby

@age     = Time.now - (5 * 24 * 3600)
@convert = false
@bw      = 'dsl2000'
@dir     = '.'
@filter  = nil
@verbose = false

@asxs  = []
@cats  = []
@items = []
@links = []

opts = GetoptLong.new(
  [ '--age',      '-a', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--convert',  '-c', GetoptLong::NO_ARGUMENT ],
  [ '--dir',      '-d', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--fast',     '-f', GetoptLong::NO_ARGUMENT ],
  [ '--match',    '-m', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--help',     '-h', GetoptLong::NO_ARGUMENT ],
  [ '--verbose',  '-v', GetoptLong::NO_ARGUMENT ],

  [ '--asx',      '-A', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--category', '-C', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--item',     '-I', GetoptLong::REQUIRED_ARGUMENT ],
  [ '--link',     '-L', GetoptLong::REQUIRED_ARGUMENT ]
)

def mt_parse_asx(link)
  data = ""
  open(link) do |s| data = s.read end
  XML::Parser.string(data).parse.root.find('Entry').first.find('Ref').first['href']
end

def mt_stream_url(link)
  data = ""
  open("#{link}?&bw=#{@bw}&pp=wmp&view=navJson") do |s| data = s.read end
  json = JSON.parse(data)
  return "INVALID:#{json['assetType']}" if json['assetType'] != "video"
  mt_parse_asx(json['assetUrl'])
end

@spinner = [ '-', '\\', '|', '/' ]

def wait(pid)
  if @verbose
    i = 0
    while Process::waitpid(pid, Process::WNOHANG) == nil
      print "\b#{@spinner[i]}"
      STDOUT.flush
      i = (i + 1) % 4
      sleep 0.05
    end

    puts "\b\b."
  else
    Process::waitpid(pid, 0)
  end
end

def dump_stream(url)
  match = /^INVALID:(.*)/.match(url)
  if match
    puts "[skip] this looks like a #{match[1]} source"
    return
  end

  basename = url.split("/").last
  outfile  = "#{@dir}/#{basename}"
  system("mkdir -p #{@dir}")

  begin File::Stat.new(outfile)
    puts "[skip] File already exists (#{outfile})" if @verbose
  rescue
    cmd = "mplayer -nolirc -really-quiet -dumpstream -dumpfile '#{outfile}' '#{url}'"
    print "[dump] #{url} -> #{outfile}  " if @verbose
    wait(fork do system(cmd) end)
    convert_stream(outfile) if @convert
  end
end

def convert_stream(file)
  print "[conv] #{file}  " if @verbose
  wait(fork do system("ffmpeg2theora '#{file}'") end)
end

opts.each do |opt, arg|
  case opt
  when '--age'
    @age = Time.now - (arg.to_i * 24 * 3600)
  when '--convert'
    @convert = true
  when '--dir'
    @dir = arg
  when '--fast'
    @bw = 'dsl1000'
  when '--match'
    @filter = Regexp.new(arg, true, 'u')
  when '--player'
    if ['wmp','qtp'].include?(arg)
      @player = arg
    else
      raise "Wrong --player, supported players: wmp, qtp"
    end
  when '--help'
    RDoc::usage
    exit 0
  when '--verbose'
    @verbose = true

  when '--asx'
    @asxs << arg
  when '--category'
    @cats << arg.to_i
  when '--item'
    @items << arg.to_i
  when '--link'
    @links << arg
  end
end

@asxs.each do |asx|
  dump_stream(mt_parse_asx(asx))
end

@cats.each do |cat|
  rss = RSS::Parser.parse("http://www.zdf.de/ZDFmediathek/content/#{cat}?view=rss")
  rss.items.each do |item|
    if item.date >= @age
      puts "[test] #{item.title}" if @verbose
      match = @filter.nil? ? true : item.title =~ @filter
      dump_stream(mt_stream_url(item.link)) if match
    end
  end
end

@items.each do |item|
  dump_stream(mt_stream_url("http://www.zdf.de/ZDFmediathek/content/#{item}"))
end

@links.each do |link|
  dump_stream(mt_stream_url(link))
end

File: mtscrape.py
#!/usr/bin/python
# vim: set fileencoding=utf8

import os, sys, subprocess, time, datetime
import re, urllib, feedparser, json
from xml.dom import minidom
from optparse import OptionParser, make_option
import logging as log

__author__ = "Benedikt Böhm"
__copyright__ = "Copyright (c) 2007-2010 Benedikt Böhm <bb@xnull.de>"
__version__ = 0,3

OPTION_LIST = [
  make_option(
    "-a", "--age",
    dest = "age",
    type = "int",
    default = 5,
    metavar = "DAYS",
    help = "maxmimum age of feed item in days"),
  make_option(
    "-c", "--convert",
    dest = "convert",
    action = "store_true",
    default = False,
    help = "convert wmv/mov to ogg/theora"),
  make_option(
    "-d", "--dir",
    dest = "dir",
    type = "string",
    default = ".",
    metavar = "PATH",
    help = "output directory"),
  make_option(
    "-f", "--fast",
    dest = "fast",
    action = "store_true",
    default = False,
    help = "stream low-quality movies (faster)"),
  make_option(
    "-m", "--match",
    dest = "match",
    type = "string",
    default = "",
    metavar = "PATTERN",
    help = "regular expression to match against feed items"),
  make_option(
    "-v", "--verbose",
    dest = "verbose",
    action = "store_true",
    default = False,
    help = "verbose output"),
  make_option(
    "-A", "--asx",
    dest = "asx",
    action = "append",
    default = [],
    metavar = "URL",
    help = "Parse ASX link/file directly"),
  make_option(
    "-C", "--category",
    dest = "cat",
    action = "append",
    default = [],
    metavar = "ID",
    help = "Parse RSS feed of mediathek category"),
  make_option(
    "-I", "--item",
    dest = "item",
    action = "append",
    default = [],
    metavar = "ID",
    help = "Parse JSON of mediathek item ID"),
  make_option(
    "-L", "--link",
    dest = "link",
    action = "append",
    default = [],
    metavar = "URL",
    help = "Parse JSON of mediathek link"),
]

def run_wait(cmd, options):
  spinner = ['-', '\\', '|', '/' ]
  proc = subprocess.Popen(cmd)

  if options.verbose:
    pos = 0
    while proc.poll() == None:
      sys.stdout.write("\b" + spinner[pos])
      sys.stdout.flush()
      time.sleep(0.05)
      pos = (pos + 1) % len(spinner)

    sys.stdout.write("\b\b.\n")
    sys.stdout.flush()
  else:
    proc.wait()

def mt_parse_asx(url):
  dom = minidom.parseString(urllib.urlopen(url).read())
  return dom.getElementsByTagName("Ref")[0].getAttribute("href")

def mt_stream_url(url, options):
  try:
    url.index('?')
    url = url + '&'
  except ValueError:
    url = url + '?'

  url = url + 'flash=off'

  log.debug('scraping URL %s' % (url,))
  dom = minidom.parseString(urllib.urlopen(url).read())

  for a in dom.getElementsByTagName('a'):
    href = a.getAttribute('href')
    if href.find('streaming.zdf.de') > 0 and href.endswith('.asx'):
      if options.fast and href.find('zdf/300/') > 0:
        return mt_parse_asx(href)
      if not options.fast and href.find('zdf/veryhigh/') > 0:
        return mt_parse_asx(href)

def convert_stream(file):
  if options.verbose:
    sys.stdout.write("[conv] %s  " % (file,))
    run_wait(["ffmpeg2theora", file], options)

def dump_stream(url, options):
  if url.startswith("INVALID:"):
    print "[skip] this looks like a %s source" % (url.replace("INVALID:",""))
    return

  basename = url.split("/")[-1]
  outfile = os.path.join(options.dir, basename)
  if not os.path.exists(options.dir):
    os.makedirs(options.dir)

  if os.path.exists(outfile):
    print "[skip] file already exists (%s)" % (outfile,)
    return

  if options.verbose:
    sys.stdout.write("[dump] %s -> %s ." % (url, outfile))
    run_wait(["mplayer", "-nolirc", "-really-quiet", "-dumpstream", "-dumpfile", outfile, url], options)

  if options.convert:
    convert_stream(outfile)

def main():
  log.basicConfig(filename='/dev/stderr', level=log.DEBUG)

  version = ".".join(map(str, __version__))
  parser = OptionParser(option_list=OPTION_LIST,version="%prog " + version)
  (options, args) = parser.parse_args()

  for asx in options.asx:
    dump_stream(mt_parse_asx(asx), options)

  for cat in options.cat:
    rss = feedparser.parse("http://www.zdf.de/ZDFmediathek/rss/%s?view=rss" % (cat,))
    age = datetime.datetime.now() - datetime.timedelta(days=options.age)
    for item in rss['items']:
      if datetime.datetime(*item.updated_parsed[:6]) >= age:
        if options.verbose:
          print "[test] " + item.title
        if re.search(options.match, item.title):
          dump_stream(mt_stream_url(item.link, options), options)

  for item in options.item:
    dump_stream(mt_stream_url("http://www.zdf.de/ZDFmediathek/beitrag/video/" + item, options), options)

  for link in options.link:
    dump_stream(mt_stream_url(link, options), options)

  sys.exit(0)

if __name__ == "__main__":
  main()
